File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/CoinChange.java
package com.example.dp;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class CoinChange {
    private static final int INF = 1_000_000_000;

    private CoinChange() {
    }

    public static Result memoized(int[] coins, int amount) {
        int value = solve(coins, amount, new HashMap<>());
        return new Result("Coin Change", "memoization", value >= INF ? -1 : value, Map.of());
    }

    private static int solve(int[] coins, int amount, Map<Integer, Integer> memo) {
        if (amount == 0) {
            return 0;
        }
        if (amount < 0) {
            return INF;
        }
        if (memo.containsKey(amount)) {
            return memo.get(amount);
        }
        int best = INF;
        for (int coin : coins) {
            best = Math.min(best, 1 + solve(coins, amount - coin, memo));
        }
        memo.put(amount, best);
        return best;
    }

    public static Result tabulated(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        int[] parent = new int[amount + 1];
        Arrays.fill(dp, INF);
        Arrays.fill(parent, -1);
        dp[0] = 0;
        for (int current = 1; current <= amount; current++) {
            for (int coin : coins) {
                if (coin <= current && dp[current - coin] + 1 < dp[current]) {
                    dp[current] = dp[current - coin] + 1;
                    parent[current] = coin;
                }
            }
        }
        List<Integer> chosen = new ArrayList<>();
        for (int current = amount; current > 0 && parent[current] != -1; current -= parent[current]) {
            chosen.add(parent[current]);
        }
        int value = dp[amount] >= INF ? -1 : dp[amount];
        return new Result("Coin Change", "tabulation", value, Map.of("coins", value == -1 ? List.of() : chosen));
    }

    public static Result spaceOptimized(int[] coins, int amount) {
        Result result = tabulated(coins, amount);
        return new Result("Coin Change", "space optimized", result.value(), result.details());
    }
}

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/coin_change.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult, require_non_negative

INF = 10**9


def memoized(coins: list[int], amount: int) -> DPResult:
    require_non_negative(amount, "amount")

    @lru_cache(maxsize=None)
    def solve(remaining: int) -> int:
        if remaining == 0:
            return 0
        if remaining < 0:
            return INF
        return 1 + min(solve(remaining - coin) for coin in coins)

    value = solve(amount)
    return DPResult("Coin Change", "memoization", -1 if value >= INF else value)


def tabulated(coins: list[int], amount: int) -> DPResult:
    require_non_negative(amount, "amount")
    dp = [INF] * (amount + 1)
    parent = [-1] * (amount + 1)
    dp[0] = 0
    for current in range(1, amount + 1):
        for coin in coins:
            if coin <= current and dp[current - coin] + 1 < dp[current]:
                dp[current] = dp[current - coin] + 1
                parent[current] = coin
    chosen: list[int] = []
    current = amount
    while current > 0 and parent[current] != -1:
        chosen.append(parent[current])
        current -= parent[current]
    value = -1 if dp[amount] >= INF else dp[amount]
    return DPResult("Coin Change", "tabulation", value, {"coins": chosen if value != -1 else []})


def space_optimized(coins: list[int], amount: int) -> DPResult:
    result = tabulated(coins, amount)
    return DPResult("Coin Change", "space optimized", result.value, result.details)
